gson/src/main/java/com/google/gson/reflect/TypeToken.java

```diff
@@ -319,8 +319,46 @@ public static <T> TypeToken<T> get(Class<T> type) {
   /**
    * Gets type literal for the parameterized type represented by applying {@code typeArguments} to
    * {@code rawType}.
+   *
+   * @throws IllegalArgumentException
+   *   If {@code rawType} is not of type {@code Class}, or if the type arguments are invalid for
+   *   the raw type
    */
   public static TypeToken<?> getParameterized(Type rawType, Type... typeArguments) {
+    $Gson$Preconditions.checkNotNull(rawType);
+    $Gson$Preconditions.checkNotNull(typeArguments);
+
+    // Perform basic validation here because this is the only public API where users
+    // can create malformed parameterized types
+    if (!(rawType instanceof Class)) {
+      // See also https://bugs.openjdk.org/browse/JDK-8250659
+      throw new IllegalArgumentException("rawType must be of type Class, but was " + rawType);
+    }
+    Class<?> rawClass = (Class<?>) rawType;
+    TypeVariable<?>[] typeVariables = rawClass.getTypeParameters();
+
+    int expectedArgsCount = typeVariables.length;
+    int actualArgsCount = typeArguments.length;
+    if (actualArgsCount != expectedArgsCount) {
+      throw new IllegalArgumentException(rawClass.getName() + " requires " + expectedArgsCount +
+          " type arguments, but got " + actualArgsCount);
+    }
+
+    for (int i = 0; i < expectedArgsCount; i++) {
+      Type typeArgument = typeArguments[i];
+      Class<?> rawTypeArgument = $Gson$Types.getRawType(typeArgument);
+      TypeVariable<?> typeVariable = typeVariables[i];
+
+      for (Type bound : typeVariable.getBounds()) {
```

Comment:
I'm wondering about this. Suppose you have `interface NumberSet<E extends Number> extends Set<E>`. Then Java will allow you to write `NumberSet<?>`, which effectively means `NumberSet<? extends Number>`. Would people expect to be able to write `TypeToken.getParameterized(NumberSet.class, wildcard)` where `wildcard` is a `Type` representing `?`? I may well be overthinking this.

Reply:
Good point, I did not consider that. Though users probably cannot easily obtain a `WildcardType`, and additionally Gson's current type resolution logic does not seem to handle wildcards that well. It seems to only use their bounds but ignores the bounds of the type variable. So users would most likely run into a `ClassCastException`.